src/Shop.ts

```diff
@@ -360,21 +360,28 @@ export const shopDescriptions = (input: ShopUpgradeNames) => {
 
 export const buyShopUpgrades = async (input: ShopUpgradeNames) => {
     let p = true;
+    let upgradeCounter = 0;
+    let totalCost = 0;
+
+    // simulate the total upgrade cost by iterating over the upgrade levels until we run out of quarks
+    do {
+        const cost = getShopCosts(input);
+        totalCost += cost;
+        player.worlds.sub(cost);
+        upgradeCounter++;
+        player.shopUpgrades[input]++;
+    } while(G['shopBuyMax'] && Number(player.worlds) >= getShopCosts(input) && player.shopUpgrades[input] < shopData[input].maxLevel);
+
     if (G['shopConfirmation']) {
-        p = await Confirm("Are you sure you'd like to purchase " + input + " for " + format(getShopCosts(input)) + " Quarks? Press 'OK' to finalize purchase.");
-    }
+        // revert the simulation by adding back the quarks and removing the shopUpgrades
+        player.worlds.add(totalCost);
+        player.shopUpgrades[input] -= upgradeCounter;
```

Comment:
I thought about only reverting if `!p` after the `Confirm`, but the user would see the decreased quarks total in the header etc. before confirming